sdk/core/azure-core/inc/azure/core/http/win_http_transport.hpp

```diff
@@ -35,41 +36,22 @@ namespace Azure { namespace Core { namespace Http {
     constexpr static size_t DefaultUploadChunkSize = 1024 * 64;
     constexpr static size_t MaximumUploadChunkSize = 1024 * 1024;
 
-    struct HandleManager final
-    {
-      Context const& m_context;
-      Request& m_request;
-      HINTERNET m_connectionHandle;
-      HINTERNET m_requestHandle;
-
-      HandleManager(Request& request, Context const& context)
-          : m_request(request), m_context(context)
-      {
-        m_connectionHandle = NULL;
-        m_requestHandle = NULL;
-      }
-
-      ~HandleManager()
+    // unique_ptr class wrapping an HINTERNET handle
+    class HINTERNET_deleter {
+    public:
+      void operator()(HINTERNET handle) noexcept
       {
-        // Close the handles and set them to null to avoid multiple calls to WinHTTP to close the
-        // handles.
-        if (m_requestHandle)
-        {
-          WinHttpCloseHandle(m_requestHandle);
-          m_requestHandle = NULL;
-        }
-
-        if (m_connectionHandle)
+        if (handle != nullptr)
         {
-          WinHttpCloseHandle(m_connectionHandle);
-          m_connectionHandle = NULL;
+          WinHttpCloseHandle(handle);
         }
       }
     };
+    using unique_HINTERNET = std::unique_ptr<void, HINTERNET_deleter>;
```

Comment:
awesome!!

I like this approach very much. However, it looks like a big change to winhttp implementation details.  I would wait on @ahsonkhan for taking a look.  And you might want to also run performance tests after these changes.

I definitely see it as an improvement.  But it has a little more complexity to be digest :)

Reply:
@ahsonkhan  and I have chatted about this earlier.